script/start-driver.sh

```diff
@@ -2,24 +2,32 @@
 
 set -eou pipefail
 
+P2P_BOOTNODES="enode://c263741b17759f3850d24d67d6c3cbc307c73e17d80c6b12a63a4792a10529d1125d00ecf7ef4c9b0dc51d28b94dfc1b8798fb524f61a1f93946748649f73b23@34.142.239.251:4001?discport=30304,enode://2f37c3affd83274b262fa2a259d32d41510dd5a48d6e916696efe7f1598cb3f905305f5989e7b6607aab50697fb2e52cb4b6904116ed67cc5fcea1e6d66ccaba@35.247.159.156:4001?discport=30304,enode://dd83dedeff622ecfca0c5edf320266506c811539a553ddd91589cdfcc9bbd74d0d620f251d8d5e1180f19a446abbdd8b6b5301e9aa6cbad35cfd9716f80f2416@34.126.90.255:4001?discport=30304"
+
+ARGS="--l1.ws "${L1_ENDPOINT_WS}" \
+    --l2.ws ws://l2_execution_engine:8546 \
+    --l1.beacon "${L1_BEACON_HTTP}" \
+    --l2.auth http://l2_execution_engine:8551 \
+    --taikoInbox "${TAIKO_INBOX_ADDRESS}" \
+    --taikoAnchor "${TAIKO_ANCHOR_ADDRESS}" \
+    --preconfirmation.whitelist ${PRECONFIRMATION_WHITELIST} \
+    --preconfirmation.serverPort 9871 \
+    --jwtSecret /data/taiko-geth/geth/jwtsecret \
+    --p2p.bootnodes ${P2P_BOOTNODES} \
+    --p2p.listen.ip 0.0.0.0 \
+    --p2p.useragent taiko \
+    --p2p.listen.tcp ${P2P_LISTEN_TCP_PORT}"
+
 if [ "$DISABLE_P2P_SYNC" = "false" ]; then
-    exec taiko-client driver \
-        --l1.ws "${L1_ENDPOINT_WS}" \
-        --l2.ws ws://l2_execution_engine:8546 \
-        --l1.beacon "${L1_BEACON_HTTP}" \
-        --l2.auth http://l2_execution_engine:8551 \
-        --taikoL1 "${TAIKO_L1_ADDRESS}" \
-        --taikoL2 "${TAIKO_L2_ADDRESS}" \
-        --jwtSecret /data/taiko-geth/geth/jwtsecret \
-        --p2p.sync \
-        --p2p.checkPointSyncUrl "${P2P_SYNC_URL}"
+    ARGS="${ARGS} --p2p.sync \
+        --p2p.checkPointSyncUrl ${P2P_SYNC_URL}"
+fi
+
+if [ -n "$PUBLIC_IP" ]; then
+    ARGS="${ARGS} --p2p.advertise.ip ${PUBLIC_IP} \
+    --p2p.advertise.udp ${P2P_UDP_PORT}"
```

Comment:
we also need --p2p.advertise.tcp set to the TCP port (default 4001).

we need --p2p.listen.udp set to the UDP port.

also --p2p.listen.tcp (set to same TCP port, default 4001 or whatever it is in the env sample) and 
we should also add
--p2p.listen.ip 0.0.0.0.

once configured please run it and confirm you get peers, you should have 6 I believe on mainnet now if discovery is working and 3 if it isn't (bc those would be just the bootnodes)

Reply:
@RogerLamTd run drive with verbosity 4, that's debug, then you can grep logs for "peering tick" and "advertise".

peering tick till show you how many peers you have. advertise will show you if you're advising the right ip and tcp/udp port

Reply:
yeah on --nat it will get IP UDP and TCP as best it can be automatically attempting.

all TCP udp and ip stuff can be left alone when using --p2p.nat except the listen 0.0.0.0

Reply:
kk i'll apply changes and re-request review, will also migrate changes to hekla bootnodes pr